text/0026-serialization.md

```diff
@@ -0,0 +1,242 @@
+
+Taking an example struct:
+
+```C#
+struct Message
+{
+    private float f;
+    private bool b;
+    private int i;
+
+    void Serialize(ref FastBufferWriter writer);
+}
+```
+
+Serialize can be implemented in two ways:
+
+```C#
+void Serialize(ref FastBufferWriter writer)
+{
+    writer.WriteSingle(f);
+    writer.WriteBool(b);
+    writer.WriteInt32(i);
+}
+```
+
+```C#
+void Serialize(ref FastBufferWriter writer)
+{
+    writer.WriteValue(this);
+}
+```
+
+The former will create more efficiently packed data in the message, and can be even further optimized by using `WriteSinglePacked()` and `WriteInt32Packed()`, but it has two downsides: first, that it involves more method calls and more instructions, making it slower; and second, that it creates a greater opportunity for the serialize and deserialize code to become misaligned, since they must contain the same operations in the same order.
+
+The latter can also be improved by optimizing the struct itself for alignment and optimizing the size of the values themselves, changing the definition to:
+
+```c#
+struct Message
+{
+    private float f;
+    private short i;
+    private bool b;
+
+    void Serialize(ref FastBufferWriter writer);
+}
+```
+
+This will naturally result in a more efficiently-packed buffer, though still not quite as efficient as using packed values.
+
+You can also use a hybrid approach if you have a few values that will need to be packed and several that won't:
+
+```C#
+struct Message
+{
+    struct Embedded
+    {
+        private byte a;
+        private byte b;
+        private byte c;
+        private byte d;
+    }
+    public Embedded embedded;
+    public float f;
+    public short i;
+
+    void Serialize(ref FastBufferWriter writer)
+    {
+        writer.WriteValue(embedded);
+        writer.WriteSinglePacked(f);
+        writer.WriteInt32Packed(i);
+    }
+}
+```
+
+This allows the four bytes of the embedded struct to be rapidly serialized as a single action, then adds the compacted data at the end, resulting in better bandwidth usage than serializing the whole struct as-is, but better performance than serializing it one byte at a time.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## FastBufferWriter and FastBufferReader
+
+FastBufferWriter and FastBufferReader are replacements for the current NetworkWriter and NetworkReader. They have much the same interface, but there are some critical differences:
+
+- FastBufferWriter and FastBufferReader are **structs**, not **classes**. This means they can be constructed and destructed without allocations if necessary.
+- FastBufferWriter and FastBufferReader both wrap `NativeArray<byte>`, allowing us to take advantage of extremely fast `Allocator.Temp` and `Allocator.TempJob` allocations to create temporary buffers.
```

Comment:
just to clarify, `FastBufferWriter`/`FastBufferReader` wouldn't create `NativeArray<byte>` instance, it'd be passed into them — so, allocator would be determined from outside, right? (such as having persistent `NativeArray<byte>`s and recycling them at different frames etc.)

Reply:
They have two constructors. One takes an existing NativeArray, the other takes a size and an allocator. In both cases you get to pick the allocation type.

Reply:
Well... Writer has two constructors... Reader has just the one.